### PR TITLE
Update matmul requires_grad checks

### PR DESCRIPTION
**Pull request by soulitzer**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #117067
* #116710
* #116523

Fixes https://github.com/pytorch/pytorch/issues/116099

### CODE REVIEW

File: test/test_linalg.py [hunk 1]
```diff
@@ -4410,6 +4410,21 @@ def test_matmul_small_brute_force_3d_Nd(self, device, dtype):
             y = make_arg(size_y, noncontiguous=nctg_y)
             self.check_single_matmul(x, y)
 
+    @dtypesIfCUDA(torch.float, torch.complex64)  # Integer matmul just supported on CPU
+    @dtypes(torch.int64, torch.float, torch.complex64)
```
Comment (lezcano) [inline test/test_linalg.py:4414]: int64 cannot require_grad=True

**Review by lezcano — APPROVED** [review]
There's a minor issue in the test, but otherwise LGTM

File: aten/src/ATen/native/LinearAlgebra.cpp [hunk 2]
```diff
@@ -1995,6 +1996,15 @@ static Tensor _matmul_impl(
 
   const bool has_out = out.defined();
 
+  if (has_out) {
+    // Usually we would rely out= kernels we decompose into to check this, but
```
Comment (albanD) [inline aten/src/ATen/native/LinearAlgebra.cpp:2000]:
```suggestion
    // Usually we would rely on the out= kernels we decompose into to check this, but
```

**Review by albanD — APPROVED** [review]

File: test/test_linalg.py (hunk 1 above)
Comment (soulitzer) [reply test/test_linalg.py:4414]: ah good catch

**Comment by malfet** [comment]
Looks like this caused a perf regression, see https://github.com/pytorch/benchmark/issues/2139

**Pull request closed without merging**